app/lib/analyzer/backend.dart

```diff
@@ -0,0 +1,152 @@
+    final Key versionKey =
+        packageKey.append(PackageVersionAnalysis, id: version);
+
+    if (analysis == null) {
+      final list = await db.lookup([versionKey]);
+      final PackageVersionAnalysis pva = list[0];
+      if (pva == null) return null;
+      analysis = pva.latestAnalysis;
+    }
+
+    // analysis was set
+    final Key analysisKey = versionKey.append(Analysis, id: analysis);
+    final List list = await db.lookup([analysisKey]);
+    return list[0];
+  }
+
+  /// Create a new [Analysis] object with the current timestamp.
+  Analysis prepareAnalysis(String package, String version) {
+    final DateTime now = new DateTime.now().toUtc();
+    final Key versionKey = db.emptyKey
+        .append(PackageAnalysis, id: package)
+        .append(PackageVersionAnalysis, id: version);
+    return new Analysis()
+      ..parentKey = versionKey
+      ..packageName = package
+      ..packageVersion = version
+      ..timestamp = now;
+  }
+
+  /// Stores the analysis, and either creates or updates its parent
+  /// [PackageAnalysis] and [PackageVersionAnalysis] rollups records.
+  Future storeAnalysis(Analysis analysis) async {
+    assert(analysis.id == null);
+    await db.commit(inserts: [analysis]);
+    assert(analysis.id != null);
+
+    // update package and version too
+    await db.withTransaction((Transaction tx) async {
+      final Key packageKey =
+          db.emptyKey.append(PackageAnalysis, id: analysis.packageName);
+      final Key packageVersionKey = packageKey.append(PackageVersionAnalysis,
+          id: analysis.packageVersion);
+      final List parents = await tx.lookup([packageKey, packageVersionKey]);
+      final PackageAnalysis package = parents[0];
+      final PackageVersionAnalysis version = parents[1];
+
+      final List<Model> inserts = [];
+      if (package == null) {
+        inserts.add(_createPackage(analysis));
+      } else {
+        if (_updatePackageLatest(package, analysis)) {
+          inserts.add(package);
+        }
+      }
+
+      if (version == null) {
+        inserts.add(_createVersion(analysis));
+      } else {
+        if (_updateVersionLatest(version, analysis)) {
+          inserts.add(version);
+        }
+      }
+
+      if (inserts.isNotEmpty) {
+        tx.queueMutations(inserts: inserts);
+        await tx.commit();
+      } else {
+        await tx.rollback();
+      }
+    });
+  }
+
+  PackageAnalysis _createPackage(Analysis analysis) {
+    final package = new PackageAnalysis()
+      ..parentKey = db.emptyKey
+      ..id = analysis.packageName
+      ..latestAnalysisKey = analysis.key;
+    _updatePackageLatest(package, analysis);
+    return package;
+  }
+
+  bool _updatePackageLatest(PackageAnalysis package, Analysis analysis) {
+    if (isNewer(analysis.semanticVersion, package.latestSemanticVersion)) {
+      // analysis version is older than latest known stable version
+      return false;
+    }
+    if (package.latestAnalysis > analysis.analysis) {
```

Comment:
That's actually not something to rely on. The numbers are not always guaranteed to be sequentially increasing.

But we also don't need that I think:

If we just created a new `Analysis` for a given `package` / `version` we simply update it's `PackageVersionAnalysis`. The only case when we don't want that is if the analysis available on the `PackageVersionAnalysis` had a newer `Analysis.analysisVersion` (e.g. was generated with a newer `package:pana` version, because we prefer newer over older reports).